jgrapht-core/src/main/java/org/jgrapht/traverse/RandomWalkIterator.java

```diff
@@ -0,0 +1,322 @@
+     *
+     * @throws IllegalArgumentException if <code>graph==null</code> or does not
+     * contain <code>startVertex</code>
+     */
+	public RandomWalkIterator(Graph<V, E> graph, V startVertex, boolean isWeighted) {
+		this(graph, startVertex, isWeighted, Double.POSITIVE_INFINITY);
+	}
+	
+	/**
+     * Creates a new iterator for the specified graph. Iteration will start at
+     * the specified start vertex. If the specified start vertex is <code>
+     * null</code>, Iteration will start at an arbitrary graph vertex.
+     * Walk is bounded by the provided number steps.
+     *
+     * @param graph the graph to be iterated.
+     * @param startVertex the vertex iteration to be started.
+     * @param isWeighted set to <code>true</code> if a weighted walk is desired.
+     * @param maxSteps number of steps before walk is exhausted.
+     *
+     * @throws IllegalArgumentException if <code>graph==null</code> or does not
+     * contain <code>startVertex</code>
+     */
+	public RandomWalkIterator(Graph<V, E> graph, V startVertex, boolean isWeighted, double maxSteps) {		
+		if (graph == null) {
+            throw new IllegalArgumentException("graph must not be null");
+        }
+		//do not cross components.
+		setCrossComponentTraversal(false);
+		this.graph = graph;
+		this.isWeighted = isWeighted;
+		this.maxSteps = maxSteps;
+		this.isDirected = graph instanceof DirectedGraph<?, ?>;
+		reusableEdgeEvent = new FlyweightEdgeEvent<V, E>(this, null);
+        reusableVertexEvent = new FlyweightVertexEvent<V>(this, null);
+        //select a random start vertex in case not provided.
+		if (startVertex == null) {
+			if (graph.vertexSet().size() > 0) {
+				currentVertex = graph.vertexSet().iterator().next();
+			}
+		} else if (graph.containsVertex(startVertex)){
+			currentVertex = startVertex;			
+		} else {
+			throw new IllegalArgumentException("graph must contain the start vertex");
+		}
+		sinkReached = false;
+	}
+
+	/**
+	 * Check if this walk is exhausted. Calling {@link #next()} on
+	 * exhausted iterator will throw {@link NoSuchElementException}.
+	 * 
+	 * @return <code>true</code>if this iterator is exhausted,
+	 * <code>false</code> otherwise.
+	 */
+	protected boolean isExhausted() {
+		return maxSteps == 0;
+	}
+	
+	/**
+     * Update data structures every time we see a vertex.
+     *
+     * @param vertex the vertex encountered
+     * @param edge the edge via which the vertex was encountered, or null if the
+     * vertex is a starting point
+     */
+	protected void encounterVertex(V vertex, E edge) {
+		maxSteps--;
+	}
+
+	/**
+     * @see java.util.Iterator#hasNext()
+     */
+	@Override
+	public boolean hasNext() {
+		return currentVertex != null && !isExhausted() &&
+				!sinkReached;
+	}
+
+	/**
+     * @see java.util.Iterator#next()
+     */
+	@Override
+	public V next() {
+		
+		if (!hasNext()) {
+			throw new NoSuchElementException();
+		}
+		
+		Set<E> potentialEdges;
+		if (isDirected) {

```

Comment:
Seems like we should move the "Specifics" logic from CrossComponentIterator up to AbstractGraphIterator so you can reuse it here.